### PR TITLE
feat: allow docs in protocol files

### LINKED ISSUE #548: Add dart docs to generated files
Adding documentation specified in yaml to generated files would be cool.

Example:
```yaml
class: Foo
documentation: "Some documentation goes here."

fields:
  # Not sure if package:yaml can extract this data.
  foo: int
  bar:
    type: String
    documentation: "This approach would required changing the yaml format. Maybe allowing a short and an expanded mode would be an option."
```

### PR DESCRIPTION
**Pull request by fischerscode**
This PR allows documenting generated files in the protocol yaml files.

Examples:
```yaml
### Some documentation goes here.
class: Foo

fields:
  ### Documentation for a field.
  foo: int
  ### Another field.
  ###
  ### Multiple lines
  bar: String
```
```yaml
### Some documentation goes here.
enum: Foo

values:
  ### Documentation for a value.
  - foo
  ### Another value
  - bar
```
The number of spaces in front of `###` is important.
(For enum values the doccumentation has to start two columns before the value name starts.)

Closes #548 

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [ ] I added new tests to check the change I am making.
- [ ] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

Not breaking.

### CODE REVIEW

**Review by yahu1031 — COMMENTED** [review]
Suggestion

File: tests/serverpod_test_client/lib/src/protocol/object_field_scopes.dart [hunk 1]
```diff
@@ -10,6 +10,7 @@ import 'package:serverpod_client/serverpod_client.dart' as _i1;
 
 class ObjectFieldScopes extends _i1.SerializableEntity {
   ObjectFieldScopes({
+    /// The database ID.
```
[the remark this answers is not in the document]
Comment (fischerscode) [reply tests/serverpod_test_client/lib/src/protocol/object_field_scopes.dart:13]: Removing them is easy, however, I think they are quite helpful when creating a new Object, since vscode shows them when hovering over the named parameter.

I guess it's open for debate :-)

[the remark this answers is not in the document]
Comment (fischerscode) [reply tests/serverpod_test_client/lib/src/protocol/object_field_scopes.dart:13]: Never mind, it seems to be caused by the `this.` syntax.
I'll remove it.

[the remark this answers is not in the document]
Comment (fischerscode) [reply tests/serverpod_test_client/lib/src/protocol/object_field_scopes.dart]: removed

File: tools/serverpod_cli/bin/util/yaml_docs.dart [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+class YamlDocumentationExtractor {
+  final List<String> lines;
+
+  YamlDocumentationExtractor(String yaml) : lines = yaml.split('\n');
+
+  List<String>? getDocumentation(SourceLocation keyStart) {
+    var documentation = <String>[];
+    var docStartExp = RegExp('^${''.padLeft(keyStart.column)}###(.*)');
+    for (var i = keyStart.line - 1; i >= 0; i--) {
```
Comment (vlidholt) [inline tools/serverpod_cli/bin/util/yaml_docs.dart:11]: I may be missing something here. But will this not continue the search upwards past non-documented fields?

E.g. will the comment for fieldA be matched for fieldB?
```yaml
class: Foo
fields:
  ### Is this matched?
  fieldA: String
  fieldB: String
```

File: tests/serverpod_test_client/lib/src/protocol/object_field_scopes.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (yahu1031) [reply tests/serverpod_test_client/lib/src/protocol/object_field_scopes.dart]: > I think they are quite helpful when creating a new Object, since vscode shows them when hovering over the named parameter.

This happens when you declare an Object/parameter.

File: tools/serverpod_cli/bin/util/yaml_docs.dart (hunk 2 above)
Comment (fischerscode) [reply tools/serverpod_cli/bin/util/yaml_docs.dart:11]: The algorithm executes line for line, starting with the line above the key.
 
`else if (line.isNotEmpty && RegExp(r'^\ *?[^#]*$').hasMatch(line))` a few lines down would break the loop when searching the docs for fieldB. So fieldB would not have a documentation.

**Review by yahu1031 — APPROVED** [review]
LGTM

File: tools/serverpod_cli/bin/util/yaml_docs.dart (hunk 2 above)
Comment (vlidholt) [reply tools/serverpod_cli/bin/util/yaml_docs.dart:11]: So hard to read those regexps.

Comment (fischerscode) [reply tools/serverpod_cli/bin/util/yaml_docs.dart:11]: I'm always using [this](https://regex101.com/r/CVky9y/1) side.

File: tools/serverpod_cli/bin/generator/class_generator_dart.dart [hunk 3]
```diff
@@ -2,7 +2,7 @@ import 'package:code_builder/code_builder.dart';
 
 import 'class_generator.dart';
 import 'config.dart';
-import 'protocol_definition.dart';
+import 'protocol_definition.dart' hide EnumValue;
```
Comment (vlidholt) [inline tools/serverpod_cli/bin/generator/class_generator_dart.dart:5]: It's somewhat confusing that the same name is used for two classes in different code files. It would be better if we could find more distinctive names as the code may be hard to follow otherwise.

Comment (fischerscode) [reply tools/serverpod_cli/bin/generator/class_generator_dart.dart:5]: `EnumValueDefinition`?

File: tools/serverpod_cli/bin/generator/class_analyzer.dart [hunk 4]
```diff
@@ -150,7 +154,8 @@ class ClassAnalyzer {
     return null;
   }
 
-  ProtocolFileDefinition? _analyzeClassFile(YamlMap documentContents) {
+  ProtocolFileDefinition? _analyzeClassFile(YamlMap documentContents,
+      List<String>? Function(SourceLocation keyStart) docsExtractor) {
```
[the remark this answers is not in the document]
Comment (fischerscode) [reply tools/serverpod_cli/bin/generator/class_analyzer.dart]: Better? af81f9d3

File: tools/serverpod_cli/bin/generator/class_generator_dart.dart (hunk 3 above)
Comment (fischerscode) [reply tools/serverpod_cli/bin/generator/class_generator_dart.dart]: Better? 08e4109b

Comment (vlidholt) [reply tools/serverpod_cli/bin/generator/class_generator_dart.dart]: Let's go with that.

**Review by vlidholt — APPROVED** [review]
